{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example Code to test the Pub_worm API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EntrezAPI imported from: /Users/dan/Code/Python/pub_worm/pub_worm/ncbi/entreze_api.py\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import inspect\n",
    "import pandas as pd\n",
    "from openpyxl import Workbook\n",
    "from openpyxl.styles import PatternFill\n",
    "from openpyxl.utils import get_column_letter\n",
    "import os\n",
    "import json\n",
    "\n",
    "# Add pub_worm directory to the Python path\n",
    "sys.path.append(\"/Users/dan/Code/Python/pub_worm\")\n",
    "\n",
    "from pub_worm.ncbi.entreze_api import EntrezAPI\n",
    "\n",
    "# Find where EntrezAPI is being load from\n",
    "module = inspect.getmodule(EntrezAPI)\n",
    "if hasattr(module, \"__file__\"):\n",
    "    file_path = module.__file__\n",
    "    print(\"EntrezAPI imported from:\", file_path)\n",
    "else:\n",
    "    print(\"Could not determine the file path.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "    {\n",
      "        \"uid\": \"10021351\",\n",
      "        \"issn\": \"0950-1991\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Ueno N\",\n",
      "        \"pmc_id\": \"\",\n",
      "        \"title\": \"Regulation of body length and male tail ray pattern formation of Caenorhabditis elegans by a member of TGF-beta family.\",\n",
      "        \"source\": \"Development\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10022905\",\n",
      "        \"issn\": \"0270-7306\",\n",
      "        \"essn\": \"1098-5549\",\n",
      "        \"last_author\": \"Kimble J\",\n",
      "        \"pmc_id\": \"PMC84011\",\n",
      "        \"title\": \"The Caenorhabditis elegans sex determination gene mog-1 encodes a member of the DEAH-Box protein family.\",\n",
      "        \"source\": \"Mol Cell Biol\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10022914\",\n",
      "        \"issn\": \"0270-7306\",\n",
      "        \"essn\": \"1098-5549\",\n",
      "        \"last_author\": \"Schlessinger J\",\n",
      "        \"pmc_id\": \"PMC84020\",\n",
      "        \"title\": \"Identification of a novel family of targets of PYK2 related to Drosophila retinal degeneration B (rdgB) protein.\",\n",
      "        \"source\": \"Mol Cell Biol\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10022975\",\n",
      "        \"issn\": \"1088-9051\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Maina CV\",\n",
      "        \"pmc_id\": \"\",\n",
      "        \"title\": \"The nuclear receptor superfamily has undergone extensive proliferation and diversification in nematodes.\",\n",
      "        \"source\": \"Genome Res\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10048487\",\n",
      "        \"issn\": \"1340-2838\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Higashitani A\",\n",
      "        \"pmc_id\": \"\",\n",
      "        \"title\": \"Characterization of a Caenorhabditis elegans recA-like gene Ce-rdh-1 involved in meiotic recombination.\",\n",
      "        \"source\": \"DNA Res\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10048958\",\n",
      "        \"issn\": \"0378-1119\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Str\\u00e5by KB\",\n",
      "        \"pmc_id\": \"\",\n",
      "        \"title\": \"Caenorhabditis elegans ZC376.5 encodes a tRNA (m2/2G(26))dimethyltransferance in which (246)arginine is important for the enzyme activity.\",\n",
      "        \"source\": \"Gene\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10049162\",\n",
      "        \"issn\": \"0378-4274\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Sedensky MM\",\n",
      "        \"pmc_id\": \"\",\n",
      "        \"title\": \"Control of anesthetic response in C. elegans.\",\n",
      "        \"source\": \"Toxicol Lett\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10049362\",\n",
      "        \"issn\": \"0890-9369\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Garriga G\",\n",
      "        \"pmc_id\": \"PMC316472\",\n",
      "        \"title\": \"The Caenorhabditis elegans gene ham-2 links Hox patterning to migration of the HSN motor neuron.\",\n",
      "        \"source\": \"Genes Dev\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10049567\",\n",
      "        \"issn\": \"0012-1606\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Fitch DH\",\n",
      "        \"pmc_id\": \"\",\n",
      "        \"title\": \"Morphogenesis of the Caenorhabditis elegans male tail tip.\",\n",
      "        \"source\": \"Dev Biol\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10049576\",\n",
      "        \"issn\": \"0012-1606\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Emmons SW\",\n",
      "        \"pmc_id\": \"\",\n",
      "        \"title\": \"Patterning of Caenorhabditis elegans posterior structures by the Abdominal-B homolog, egl-5.\",\n",
      "        \"source\": \"Dev Biol\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10051671\",\n",
      "        \"issn\": \"0027-8424\",\n",
      "        \"essn\": \"1091-6490\",\n",
      "        \"last_author\": \"Johnson CD\",\n",
      "        \"pmc_id\": \"PMC26813\",\n",
      "        \"title\": \"Reverse genetic analysis of Caenorhabditis elegans presenilins reveals redundant but unequal roles for sel-12 and hop-1 in Notch-pathway signaling.\",\n",
      "        \"source\": \"Proc Natl Acad Sci U S A\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10051850\",\n",
      "        \"issn\": \"1079-5006\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Ishii N\",\n",
      "        \"pmc_id\": \"\",\n",
      "        \"title\": \"Protein carbonyl accumulation in aging dauer formation-defective (daf) mutants of Caenorhabditis elegans.\",\n",
      "        \"source\": \"J Gerontol A Biol Sci Med Sci\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10064800\",\n",
      "        \"issn\": \"0006-8993\",\n",
      "        \"essn\": \"\",\n",
      "        \"last_author\": \"Ohshima Y\",\n",
      "        \"pmc_id\": \"\",\n",
      "        \"title\": \"Functional reconstitution of a heteromeric cyclic nucleotide-gated channel of Caenorhabditis elegans in cultured cells.\",\n",
      "        \"source\": \"Brain Res\"\n",
      "    },\n",
      "    {\n",
      "        \"uid\": \"10066248\",\n",
      "        \"issn\": \"0270-6474\",\n",
      "        \"essn\": \"1529-2401\",\n",
      "        \"last_author\": \"Kaplan JM\",\n",
      "        \"pmc_id\": \"PMC6782580\",\n",
      "        \"title\": \"Distinct signaling pathways mediate touch and osmosensory responses in a polymodal sensory neuron.\",\n",
      "        \"source\": \"J Neurosci\"\n",
      "    }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "# Example call to ePost and eSummary\n",
    "data = [\"10021351\", \"10022905\", \"10022914\", \"10022975\", \"10048487\", \"10048958\", \"10049162\", \"10049362\", \"10049567\", \"10049576\", \"10051671\", \"10051850\", \"10064800\", \"10066248\"]\n",
    "\n",
    "ncbi_api = EntrezAPI()\n",
    "entreze_epost_result = ncbi_api.entreze_epost(data)\n",
    "\n",
    "if 'WebEnv' in entreze_epost_result:\n",
    "    entreze_summaries_result = ncbi_api.entreze_pmid_summaries(entreze_epost_result)\n",
    "    pretty_data = json.dumps(entreze_summaries_result, indent=4)\n",
    "    print(pretty_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Emple code to Get full Articles from entrez/eutils\n",
    "import requests\n",
    "import time\n",
    "\n",
    "def fetch_full_article(pmid):\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    params = {\n",
    "        \"db\": \"pmc\",\n",
    "        \"id\": pmid,\n",
    "        \"retmode\": \"xml\"\n",
    "    }\n",
    "\n",
    "    response = requests.get(base_url, params=params)\n",
    "    time.sleep(2)\n",
    "\n",
    "    if response.status_code == 200:\n",
    "        return response.text\n",
    "    else:\n",
    "        print(f\"Failed to fetch article for PMID {pmid}. Status code: {response.status_code}\")\n",
    "        return None\n",
    "\n",
    "# Example usage\n",
    "pmid = \"10022975\"  # Replace with the PMID of the article you want to fetch\n",
    "pmids = [\"10021351\", \"10022905\", \"10022914\", \"10022975\", \"10048487\", \"10048958\", \"10049162\", \"10049362\", \"10049567\", \"10049576\", \"10051671\", \"10051850\", \"10064800\", \"10066248\"]\n",
    "for pmid in pmids:\n",
    "    article_xml = fetch_full_article(pmid)\n",
    "    if article_xml:\n",
    "        with open(f\"./output/article_{pmid}.xml\", \"w\") as file:\n",
    "            file.write(article_xml)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BASH_SILENCE_DEPRECATION_WARNING: 1\n",
      "COMMAND_MODE: unix2003\n",
      "CONDA_DEFAULT_ENV: dan-dev\n",
      "CONDA_EXE: /Users/dan/miniforge3/bin/conda\n",
      "CONDA_PREFIX: /Users/dan/miniforge3/envs/dan-dev\n",
      "CONDA_PROMPT_MODIFIER: (dan-dev) \n",
      "CONDA_PYTHON_EXE: /Users/dan/miniforge3/bin/python\n",
      "CONDA_SHLVL: 2\n",
      "DISPLAY: /private/tmp/com.apple.launchd.LMlyQh3rCw/org.xquartz:0\n",
      "DOCKERHUB_API_TOKEN: dckr_pat_sxz8u7vSdIIYBtD502f9ihe3izU\n",
      "GITHUB_ACCESS_TOKEN: hp_VpgjfpywnL2WYUcyhQJh4xs2PbBNoe4QkD82\n",
      "GITHUB_USERNAME: DanH-UMassMed\n",
      "GROOVY_HOME: /Users/dan/.sdkman/candidates/groovy/current\n",
      "HOME: /Users/dan\n",
      "HOMEBREW_CELLAR: /opt/homebrew/Cellar\n",
      "HOMEBREW_PREFIX: /opt/homebrew\n",
      "HOMEBREW_REPOSITORY: /opt/homebrew\n",
      "HUGGINGFACEHUB_API_TOKEN: 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\n",
      "INFOPATH: /opt/homebrew/share/info:\n",
      "JAVA_HOME: /Users/dan/.sdkman/candidates/java/current\n",
      "LOGNAME: dan\n",
      "MANPATH: /opt/homebrew/share/man:\n",
      "MAVEN_HOME: /opt/apache-maven-3.8.5\n",
      "MONGO_PATH: /Users/dan/mongo_database/mongodb\n",
      "MallocNanoZone: 0\n",
      "NCBI_API_KEY: 5a1e23e51e6bf572435b326f1452a339ce08\n",
      "NVM_BIN: /Users/dan/.nvm/versions/node/v16.16.0/bin\n",
      "NVM_DIR: /Users/dan/.nvm\n",
      "NVM_INC: /Users/dan/.nvm/versions/node/v16.16.0/include/node\n",
      "OPENAI_API_KEY: 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\n",
      "ORCID_ACCESS_TOKEN: c0407918-2b0c-4159-9188-c64719764024\n",
      "ORCID_REFRESH_TOKEN: ac3208c6-668a-4403-b3e1-fb5c6438cb94\n",
      "ORIGINAL_XDG_CURRENT_DESKTOP: undefined\n",
      "PATH: /Users/dan/miniforge3/envs/dan-dev/bin:/Users/dan/.sdkman/candidates/java/current/bin:/Users/dan/.sdkman/candidates/groovy/current/bin:/Users/dan/Applications/google-cloud-sdk/bin:/Users/dan/miniforge3/envs/dan-dev/bin:/Users/dan/miniforge3/condabin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/dan/.nvm/versions/node/v16.16.0/bin:/Users/dan/bin:.:/opt/apache-maven-3.8.5/bin:/usr/local/mysql/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/opt/X11/bin:/Library/Apple/usr/bin:/Library/TeX/texbin:/Applications/VMware Fusion.app/Contents/Public:/Users/dan/mongo_database/mongodb/bin\n",
      "PS1: (base) \\w $\n",
      "PWD: /\n",
      "PYPI_API_TOKEN: pypi-AgEIcHlwaS5vcmcCJGJhNzRjMDUwLWEzYzUtNDhiMC04MmFiLWQ0ZjA5NTEzZjY2ZgACKlszLCI3MWU5Y2I0Yi0zNGI2LTRhNzItYTkwMi05ZmM0YzY4NTIyMzEiXQAABiBixmwfpZ3caURsXd1_mVjtQCIrMNdFYfHtxkC0EoaKxw\n",
      "PYSPARK_DRIVER_PYTHON: jupyter\n",
      "PYSPARK_DRIVER_PYTHON_OPTS: notebook\n",
      "PYSPARK_PYTHON: python3\n",
      "SDKMAN_CANDIDATES_API: https://api.sdkman.io/2\n",
      "SDKMAN_CANDIDATES_DIR: /Users/dan/.sdkman/candidates\n",
      "SDKMAN_DIR: /Users/dan/.sdkman\n",
      "SDKMAN_PLATFORM: darwinarm64\n",
      "SERP_API_KEY: 8a3d217898841e4958cc437d577d5738c42af365cf1baa9818a4f8676c106064\n",
      "SHELL: /bin/bash\n",
      "SHLVL: 3\n",
      "SPARK_HOME: /Users/dan/miniforge3/envs/pyspark/lib/python3.9/site-packages/pyspark\n",
      "SSH_AUTH_SOCK: /private/tmp/com.apple.launchd.T8wiZoWK8e/Listeners\n",
      "TAVILY_API_KEY: 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\n",
      "TMPDIR: /var/folders/jq/7cv336vx66v_c3vj7vwrpt6c0000gn/T/\n",
      "USER: dan\n",
      "VSCODE_AMD_ENTRYPOINT: vs/workbench/api/node/extensionHostProcess\n",
      "VSCODE_CODE_CACHE_PATH: /Users/dan/Library/Application Support/Code/CachedData/863d2581ecda6849923a2118d93a088b0745d9d6\n",
      "VSCODE_CRASH_REPORTER_PROCESS_TYPE: extensionHost\n",
      "VSCODE_CWD: /\n",
      "VSCODE_HANDLES_UNCAUGHT_ERRORS: true\n",
      "VSCODE_IPC_HOOK: /Users/dan/Library/Application Support/Code/1.87-main.sock\n",
      "VSCODE_NLS_CONFIG: {\"locale\":\"en-us\",\"osLocale\":\"en-us\",\"availableLanguages\":{},\"_languagePackSupport\":true}\n",
      "VSCODE_PID: 87912\n",
      "XML_CATALOG_FILES: file:///Users/dan/miniforge3/etc/xml/catalog file:///etc/xml/catalog\n",
      "XPC_FLAGS: 0x0\n",
      "XPC_SERVICE_NAME: 0\n",
      "_: /Users/dan/miniforge3/envs/dan-dev/bin/python\n",
      "__CFBundleIdentifier: com.microsoft.VSCode\n",
      "__CF_USER_TEXT_ENCODING: 0x1F5:0x0:0x0\n",
      "ELECTRON_RUN_AS_NODE: 1\n",
      "APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL: 1\n",
      "PYTHONUNBUFFERED: 1\n",
      "PYTHONIOENCODING: utf-8\n",
      "_CE_CONDA: \n",
      "CONDA_PREFIX_1: /Users/dan/miniforge3\n",
      "CONDA_ROOT: /Users/dan/miniforge3\n",
      "_CE_M: \n",
      "LC_CTYPE: UTF-8\n",
      "PYDEVD_IPYTHON_COMPATIBLE_DEBUGGING: 1\n",
      "PYDEVD_USE_FRAME_EVAL: NO\n",
      "TERM: xterm-color\n",
      "CLICOLOR: 1\n",
      "FORCE_COLOR: 1\n",
      "CLICOLOR_FORCE: 1\n",
      "PAGER: cat\n",
      "GIT_PAGER: cat\n",
      "MPLBACKEND: module://matplotlib_inline.backend_inline\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "env_vars = os.environ\n",
    "\n",
    "# Print each environment variable\n",
    "for key, value in env_vars.items():\n",
    "    print(f\"{key}: {value}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dan-dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
